srml/consensus/src/lib.rs

```diff
@@ -292,34 +288,44 @@ impl<T: Trait> Module<T> {
 }
 
 impl<T: Trait> ProvideInherent for Module<T> {
-	type Inherent = <T::InherentOfflineReport as InherentOfflineReport>::Inherent;
 	type Call = Call<T>;
-
-	fn create_inherent_extrinsics(data: Self::Inherent) -> Vec<(u32, Self::Call)> {
-		if <T::InherentOfflineReport as InherentOfflineReport>::is_empty(&data) {
-			vec![]
+	type Error = MakeFatalError<RuntimeString>;
+	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+		if let Ok(Some(data)) =
+			data.get_data::<<T::InherentOfflineReport as InherentOfflineReport>::Inherent>(
+				&INHERENT_IDENTIFIER
+			)
+		{
+			if <T::InherentOfflineReport as InherentOfflineReport>::is_empty(&data) {
+				None
+			} else {
+				Some(Call::note_offline(data))
+			}
 		} else {
-			vec![(T::NOTE_OFFLINE_POSITION, Call::note_offline(data))]
+			None
 		}
 	}
 
-	fn check_inherent<Block: BlockT, F: Fn(&Block::Extrinsic) -> Option<&Self::Call>>(
-		block: &Block, expected: Self::Inherent, extract_function: &F
-	) -> result::Result<(), CheckInherentError> {
-		let noted_offline = block
-			.extrinsics()
-			.get(T::NOTE_OFFLINE_POSITION as usize)
-			.and_then(|xt| match extract_function(&xt) {
-				Some(Call::note_offline(ref x)) => Some(x),
-				_ => None,
-			});
-
-		// REVIEW: perhaps we should be passing a `None` to check_inherent.
-		if let Some(noted_offline) = noted_offline {
-			<T::InherentOfflineReport as InherentOfflineReport>::check_inherent(&noted_offline, &expected)
-				.map_err(|e| CheckInherentError::Other(e.into()))?;
-		}
-
-		Ok(())
+	fn check_inherent(call: &Self::Call, data: &InherentData) -> Result<(), Self::Error> {
+		let offline = match call {
+			Call::note_offline(ref offline) => offline,
+			_ => return Ok(()),
+		};
+
+		let expected =
+			match data.get_data::<<T::InherentOfflineReport as InherentOfflineReport>::Inherent>(
```

Comment:
Can't we do this more directly (and less weirdly wrapped) in something like this?
```suggestion
let expected = data.get_data::<<T::InherentOfflineReport as InherentOfflineReport>::Inherent(&INHERENT_IDENTIFIER)?
    .ok_or(Err(RuntimeString::from("No `offline_report` found in the inherent data!"))?;
```